### DISCUSSION TITLE
How to debug "Fatal error while logging another fatal error."

### DISCUSSION THREAD

**Discussion by nietras** [dotnet/runtime#83077] [Q&A]
Hi, we have being seeing `Fatal error while logging another fatal error.` messages written to console in a WPF application we have ported from .NET Framework to .NET 6. This happens at the end of application closing. 

This message appears to have been added in https://github.com/dotnet/runtime/pull/46561 ("Add recursion / thread interaction guard in LogInfoForFatalError") but how can one debug find the cause of this issue?

If using native debugger we get something like below. Note the error happens after `Main() { ... return 0; }`.
![image](https://user-images.githubusercontent.com/10798831/223374785-c08b863b-8e0d-44ae-ab17-6f86a44923fd.png)

cc: @jkotas

**Comment by jkotas** [comment]
mscorwks.dll is .NET Framework 2.0. Is it expected that your app is loading both .NET 6 and .NET Framework 2.0?

Loading multiple .NET runtimes into one process is not fully supported. In particular, Visual Studio debugger does not work well when multiple .NET runtimes are loaded into one process. If it is expected for both .NET 6 and .NET Framework 2.0 to be loaded into your app, you may need to switch to windbg to debug this crash.

**Comment by nietras** [comment]
@jkotas thanks for your quick reply.

Yes, unfortunately both .NET Framework 2.0 and .NET 6 is expected to load due to a mixed mode assembly dependency targeting .NET Fx 3.5 (I believe), as also discussed in https://github.com/dotnet/runtime/issues/80422 ("Q: Type 'IconBufferEventArgs' has a ComVisible(false) parent 'System.EventArgs' in its hierarchy, therefore QueryInterface calls for IDispatch or class interfaces are disallowed").

 Okay, I will try WinDBG, any pointers what one should look out for?

**Reply by jkotas** [reply]
The first thing to find out is stacktrace of the crash.

**Reply by nietras** [reply]
This is what I get when using WinDbg 1.2210.3001.0. Debug output has the following:
```
(44d4.566c): Unknown exception - code 0000071a (first chance)
(44d4.5ae4): CLR exception - code e0434352 (first chance)
(44d4.5ad4): CLR exception - code e0434352 (first chance)
(44d4.5aa0): CLR exception - code e0434352 (first chance)
ModLoad: 00007ff8`31d50000 00007ff8`31d6b000   C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App\6.0.14\vcruntime140_cor3.dll
ntdll!NtRemoveIoCompletion+0x14:
00007ff8`a7dcd1e4 c3              ret
```
And only one thread is active when "Fatal error while logging another fatal error." occurs/WinDbg stops.
![image](https://user-images.githubusercontent.com/10798831/223670821-8a7d10b0-3cf2-40e4-8e6b-dbf562f37298.png)
With the stack:
```
[0x0]   ntdll!NtRemoveIoCompletion + 0x14   
[0x1]   KERNELBASE!GetQueuedCompletionStatus + 0x4f   
[0x2]   System_Private_CoreLib + 0x1a8051   
[0x3]   System_Private_CoreLib!System.Threading.LowLevelLifoSemaphore.WaitForSignal + 0x32   
[0x4]   System_Private_CoreLib!System.Threading.LowLevelLifoSemaphore.Wait + 0x176   
[0x5]   System_Private_CoreLib!System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart + 0x193   
[0x6]   coreclr!CallDescrWorkerInternal + 0x83   
[0x7]   coreclr!DispatchCallSimple + 0x80   
[0x8]   coreclr!ThreadNative::KickOffThread_Worker + 0x63   
[0x9]   coreclr!ManagedThreadBase_DispatchInner + 0xd   
[0xa]   coreclr!ManagedThreadBase_DispatchMiddle + 0x85   
[0xb]   coreclr!ManagedThreadBase_DispatchOuter + 0xae   
[0xc]   coreclr!ManagedThreadBase_FullTransition + 0x2d   
[0xd]   coreclr!ManagedThreadBase::KickOff + 0x2d   
[0xe]   coreclr!ThreadNative::KickOffThread + 0x79   
[0xf]   KERNEL32!BaseThreadInitThunk + 0x14   
[0x10]   ntdll!RtlUserThreadStart + 0x21   
```

**Reply by nietras** [reply]
If I run `!pe` I get 🤔
```
0:096> !pe
c0020001 Exception in sos.pe debugger extension.
      PC: 00007ff8`a55bcd29  VA: 0000020e`6ca20000  R/W: 8007042b  Parameter: 00000000`00000000
```

**Reply by jkotas** [reply]
I think this stacktrace is too late. We need to find the stacktrace when `Fatal error while logging another fatal error` gets printed. Could you please try to set a breakpoint at the line that prints `Fatal error while logging another fatal error`?
